### ISSUE TITLE
Compiler API Suggestion: Introduce an API that facilitates knowing a record's primary constructor

### ISSUE THREAD

**Issue by Youssef1313** [dotnet/roslyn#53092]
Given an `ITypeSymbol` with `IsRecord` equals to true, there is no way to know its primary constructor symbol without having to go back to syntax.

### Proposal 1

Add `PrimaryConstructor` property to `ITypeSymbol`, which would be `null` if `IsRecord` is false, and gives the primary constructor method symbol if `IsRecord` is true.

### Proposal 2

Add `MethodKind.RecordPrimaryConstructor` to the `MethodKind` enum. That way, one could say `recordType.InstanceConstructors.FirstOrDefault(c => c.MethodKind == MethodKind.RecordPrimaryConstructor)`.

**Comment by RikkiGibson** [comment]
Identifying primary constructors from source and from metadata is a requirement before we can do #44763.

**Comment by Youssef1313** [comment]
> Identifying primary constructors from source and from metadata is a requirement before we can do #44763.

I believe this also can help @davidwengier for some EnC work too (not a requirement - but can make things easier and neater)

@RikkiGibson Would it be okay if I worked on this and opened a PR or should the API

**Comment by RikkiGibson** [comment]
I think before we work on adding the API we need to get the team to decide if we have or want a mechanism to identify primary constructors from metadata, and based on the answer to that, whether we want this API, and how it should behave.

**Comment by Youssef1313** [comment]
Closing in favor of a more detailed proposal https://github.com/dotnet/roslyn/issues/54286.

Thanks @CyrusNajmabadi

**Comment by RikkiGibson** [comment]
It wasn't clear to me how adding the AssociatedSymbol reference from the constructor parameter to the synthesized property solves the problem of knowing whether a constructor is the primary constructor.

**Comment by Youssef1313** [comment]
The other proposal doesn't solve this particular problem *directly*, but it should be achievable by looping through the first parameter of every ctor checking for a non-null IParameterSymbol.AssociatedSymbol.

**Comment by RikkiGibson** [comment]
It doesn't seem like it will help in all scenarios because it's possible to declare a primary constructor which doesn't synthesize any properties e.g.

```cs
record Rec(string X)
{
    public string X { get; init; } = X;
}
```

**Comment by Youssef1313** [comment]
Thanks for pointing that out @RikkiGibson. I missed this. (I actually forgot to close when I meant to do that, so let's keep that open).

**Comment by RikkiGibson** [comment]
Perhaps the IMethodSymbol for the primary constructor should have an AssociatedSymbol which is the record, similar to how (IIRC) we're making it so a record synthesized property has an AssociatedSymbol which is the primary constructor parameter.
